### PR TITLE
09-order-item-history-management

### PR DESCRIPTION
**Pull request by andrey-kiyas**
Added:
- order
- item
- history
- management

### CODE REVIEW

File: src/main/java/com/onlinebookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return orderService.findAllOrderItems(orderId);
+    }
+
+    @Operation(summary = "Retrieve a specific OrderItem within an order",
+            description = "Retrieve a specific OrderItem within an order")
+    @GetMapping("/{orderId}/items/{itemId}")
+    public OrderItemResponseDto findOrderItemById(@PathVariable Long orderId,
+                                                  @PathVariable Long itemId) {
+        return orderService.findOrderItemById(orderId, itemId);
+    }
```
Comment (aaananas) [inline src/main/java/com/onlinebookstore/controller/OrderController.java]: lets add preAuthorize

File: src/main/java/com/onlinebookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+@Tag(name = "Order management", description = "Endpoints for managing users orders")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+
+    @Operation(summary = "Place an order", description = "Place an order")
+    @PostMapping()
```
Comment (aaananas) [inline src/main/java/com/onlinebookstore/controller/OrderController.java]:
```suggestion
    @PostMapping
```

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(cartItem -> {
+                    OrderItem orderItem = new OrderItem();
+                    orderItem.setBook(cartItem.getBook());
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    orderItem.setOrder(order);
+                    orderItem.setPrice(cartItem.getBook().getPrice()
+                            .multiply(new BigDecimal(cartItem.getQuantity())));
+                    return orderItem;
+                })
```
Comment (aaananas) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]: lets use a mapper for this

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/com/onlinebookstore/controller/OrderController.java (hunk 1 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/OrderController.java]: I don't know why, but I decided that these endpoints should be available to both the user and the admin. Only the user is specified in the job. Corrected, thanks!

File: src/main/java/com/onlinebookstore/controller/OrderController.java (hunk 2 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/OrderController.java]: I did something here but I don’t remember that I didn’t remove the brackets. Corrected, thanks!

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java (hunk 3 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]: I could not get the mapper to process this part correctly, so I moved this part into a separate method. How can I do it the right way?

File: src/main/java/com/onlinebookstore/model/OrderItem.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = false")
+@Table(name = "order_items")
+public class OrderItem {
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/model/OrderItem.java:24]: add equals/hashCode methods

File: src/main/java/com/onlinebookstore/repository/OrderRepository.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    @Query("SELECT o FROM Order o LEFT JOIN FETCH o.orderItems LEFT JOIN FETCH o.user")
+    List<Order> findAllOrders(Pageable pageable);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/repository/OrderRepository.java]: why do you need this method?

File: src/main/java/com/onlinebookstore/repository/OrderRepository.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    @Modifying
+    @Transactional
+    @Query("UPDATE Order o SET o.status = :status WHERE o.id = :orderId")
+    void updateOrderByStatus(Long orderId, Order.Status status);
+
+    @Query("SELECT o FROM Order o LEFT JOIN FETCH o.orderItems WHERE o.id = :id")
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/repository/OrderRepository.java]:
```suggestion
    @EntityGraph(attributePaths = {"orderItems", "orderItems.book"}))
```

File: src/main/java/com/onlinebookstore/controller/OrderController.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                                   @RequestBody @Valid OrderRequestDto orderRequestDto) {
+        User user = (User) authentication.getPrincipal();
+        return orderService.create(user.getId(), orderRequestDto);
+    }
+
+    @Operation(summary = "Retrieve user's order history",
+            description = "Retrieve user's order history")
+    @GetMapping
+    public List<OrderResponseDto> findAll(Pageable pageable) {
+        return orderService.findAllOrders(pageable);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/controller/OrderController.java]: userId?

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    private Set<OrderItem> getOrderItemsFromShoppingCart(ShoppingCart shoppingCart, Order order) {
+        return shoppingCart.getCartItems().stream()
+                .map(cartItem -> {
+                    OrderItem orderItem = new OrderItem();
+                    orderItem.setBook(cartItem.getBook());
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    orderItem.setOrder(order);
+                    orderItem.setPrice(cartItem.getBook().getPrice()
+                            .multiply(new BigDecimal(cartItem.getQuantity())));
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java:85]:
```suggestion
                    orderItem.setPrice(cartItem.getBook().getPrice());
```

File: src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        BigDecimal total = orderItems.stream()
+                .map(orderItem -> orderItem.getBook().getPrice()
+                        .multiply(new BigDecimal(orderItem.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+        order.setTotal(total);
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public List<OrderResponseDto> findAllOrders(Pageable pageable) {
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
    public List<OrderResponseDto> findAllOrders(Long userId, Pageable pageable) {
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/onlinebookstore/repository/OrderRepository.java (hunk 6 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/repository/OrderRepository.java]: This works: "@EntityGraph(attributePaths = "orderItems")"

File: src/main/java/com/onlinebookstore/repository/OrderRepository.java (hunk 5 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/repository/OrderRepository.java]: Sorry, corrected to search by user

File: src/main/java/com/onlinebookstore/model/OrderItem.java (hunk 4 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/model/OrderItem.java:24]: Done! )

**Pull request merged**